src/lib/users/ssh_authorized_keyring.rb

```diff
@@ -0,0 +1,213 @@
+        end
+        owner, group = perms_from(home)
+        if create_ssh_dir(home, owner, group, SSH_DIR_PERMS)
+          write_file(home, owner, group, AUTHORIZED_KEYS_PERMS)
+        else
+          log.error("SSH directory does not exist and could not be created: giving up")
+          return false
+        end
+      end
+
+      private
+
+      # @return [String] Relative path to the SSH directory inside users' home
+      SSH_DIR = ".ssh".freeze
+      # @return [String] Authorized keys file name
+      AUTHORIZED_KEYS_FILE = "authorized_keys".freeze
+      # @return [String] Permissions to be set on SSH_DIR directory
+      SSH_DIR_PERMS = "0700".freeze
+      # @return [String] Permissions to be set on `authorized_keys` file
+      AUTHORIZED_KEYS_PERMS = "0600".freeze
+
+      # Determine the path to the user's SSH directory
+      #
+      # @param home [String] Home directory
+      # @return [String] Path to the user's SSH directory
+      #
+      # @see SSH_DIR
+      def ssh_dir_path(home)
+        File.join(home, SSH_DIR)
+      end
+
+      # Determine the path to the user's authorized keys file
+      #
+      # @param home [String] Home directory
+      # @return [String] Path to authorized keys file
+      #
+      # @see SSH_DIR
+      # @see AUTHORIZED_KEYS_FILE
+      #
+      # @see #ssh_dir_path
+      def authorized_keys_path(home)
+        File.join(ssh_dir_path(home), AUTHORIZED_KEYS_FILE)
+      end
+
+      # Find or creates the SSH directory
+      #
+      # This method sets up the SSH directory (usually .ssh). Although only 1
+      # level is needed (as SSH directory lives under $HOME/.ssh), this code
+      # should support changing SSH_DIR to something like `.config/ssh`.
+      #
+      # @param home  [String] Home directory where SSH directory must be created
+      # @param owner [Fixnum] Owner's UID
+      # @param group [Fixnum] Owner's GID
+      # @param perms [String] Permissions (in form "0700")
+      # @return [String] Returns the path to the first created directory
+      def create_ssh_dir(home, owner, group, perms)
+        ssh_dir = ssh_dir_path(home)
+        return ssh_dir if FileUtils::Exists(ssh_dir)
+        ret = SCR.Execute(Path.new(".target.mkdir"), ssh_dir)
+        log.info("Creating SSH directory: #{ret}")
+        return false unless ret
+        set_owner_and_perms(ssh_dir, owner, group, perms)
+      end
+
+      # Write authorized keys file
+      #
+      # @param path  [String] Path to file/directory
+      # @param owner [Fixnum] Owner's UID
+      # @param group [Fixnum] Owner's GID
+      # @param perms [String] Permissions (in form "0700")
+      def write_file(home, owner, group, perms)
+        path = authorized_keys_path(home)
+        file = SSHAuthorizedKeysFile.new(path)
+        file.keys = keys[home]
+        log.info "Writing #{keys[home].size} keys in #{path}"
+        file.save && set_owner_and_perms(path, owner, group, perms)
```

Comment:
This might be also a possible issue, the file is written and **then** the permissions are set. There is a theoretic chance that someone can read the data in the meantime. (Well, it depends on the default `umask`, but anyway...)

On the other hand we are writing the *public* SSH key so you cannot abuse the file even if you read it.

So try to fix that if possible, if not then at least document it.